### PR TITLE
Add Glue data source

### PR DESCRIPTION
**Pull request by MariiaRa**
Added dependencies for Athena driver and AWS Glue client, created class for representing access logic to Athena and added methods for getting table location, setting table location and recovering partitions

### CODE REVIEW

File: bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/connectors/GlueConnector.scala [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  /**
+    * @param tableName - name of table
+    * @return - list of Partiton objects which contain metadata
+    */
+  private def getPartitions(tableName: String): List[Partition] = {
+    val request = new GetPartitionsRequest
+    request.setDatabaseName(config.database)
+    request.setTableName(tableName)
+    glueClient.getPartitions(request).getPartitions.toList
```
Comment (NikolayKhoroshevsky) [inline bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/connectors/GlueConnector.scala]: It will not load all partitions from Athena. 
To load all the partitions you need to use `tokens`  and `segments`.

File: bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/metrics/GlueMetricFactory.scala [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+import com.aol.one.dwh.bandarlog.connectors.{GlueConnector, JdbcConnector}
+import com.aol.one.dwh.bandarlog.metrics.BaseMetrics.{IN, LAG, OUT}
+import com.aol.one.dwh.bandarlog.metrics.Metrics.REALTIME_LAG
+import com.aol.one.dwh.bandarlog.providers.{CurrentTimestampProvider, GlueTimestampProvider, SqlLagProvider, SqlTimestampProvider}
+import com.aol.one.dwh.infra.config.{ConnectorConfig, TableColumn, Tag}
+import com.aol.one.dwh.infra.sql.MaxValuesQuery
+import com.aol.one.dwh.infra.sql.pool.ConnectionPoolHolder
+
+class GlueMetricFactory(connectionPoolHolder: ConnectionPoolHolder) {
```
Comment (alexey-lipodat) [inline bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/metrics/GlueMetricFactory.scala]: I guess we can do this in another way, without separate MetricFactory. 
Let's discuss this.

File: bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/BandarlogsFactory.scala [hunk 3]
```diff
@@ -56,10 +56,24 @@ class BandarlogsFactory(mainConfig: Config) extends LogTrait with ExceptionPrint
     bandarlogConf.getBandarlogType match {
       case "kafka" => kafkaMetricProviders(bandarlogConf)
       case "sql" => sqlMetricProviders(bandarlogConf, connectionPoolHolder)
+      case "glue" => glueMetricProvider(bandarlogConf, connectionPoolHolder)
       case t => throw new IllegalArgumentException(s"Unsupported bandarlog type:[$t]")
     }
   }
 
+  private def glueMetricProvider(bandarlogConf: Config, connectionPoolHolder: ConnectionPoolHolder) = {
```
Comment (alexey-lipodat) [inline bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/BandarlogsFactory.scala]: The same here, we don't need the separate provider.
Actually, this provider is the almost the same as the SQL provider, the one difference is - connector.

**Comment by alexey-lipodat** [comment]
You are done a good job, but let's discuss these changes.
I'm pretty sure that we can find the more simple way to do this.

File: bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/connectors/GlueConnector.scala (hunk 1 above)
Comment (MariiaRa) [reply bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/connectors/GlueConnector.scala]: Hello! Thanks for the reply :) Can you, please, explain if I have understood Segment thing correctly? We have SegmentNumber and TotalSegments, and we can set these value. If I set TotalSegments=1, then request will be executed over the whole table. But, for example, I can also set the number of segments = 4, then SegmentNumber values will range from 0 through 3.
If this is the case, then do we need to get all partitions from the table, or just from the last segment? Is my understanding correct? 
Thanks again!

**Comment by alexey-kotlyar** [comment]
please fix stylistic error mentioned in travis log

```[error] /home/travis/build/OneByAol/bandar-log/bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/metrics/SqlMetricFactory.scala: File must end with newline character```

**Comment by alexey-lipodat** [comment]
In case of any questions - you can always ping me in Slack/Skype.

File: bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/connectors/GlueConnector.scala [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  /**
+    * Calculates max value in Partition list
+    *
+    * @param tableName   - table name
+    * @param tableColumn - column name
+    * @param list        - list of Partitions
+    * @return - max value in Partition list
+    */
+
+  private def temporaryMax(tableName: String, tableColumn: String, list: List[Partition]): Long = {
```
[the remark this answers is not in the document]
Comment (MariiaRa) [reply bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/connectors/GlueConnector.scala]: Here temporary max is the max value of batchId for _one_ request. In general, logic behind getting all partitions is following: in order to be sure we load all partitions from Athena, first we split the table into segments (Segment defines a non-overlapping region of a table's partitions, allowing multiple requests to be executed in parallel and max number of segments could be 10); then we set value withMaxResults( fetchSize ), fetchSize  is the maximum number of partitions to return in a single response. Making the first request to load partitions we get token, which can be used for the second request and so on; we are getting tokens until we have some partitions to load, otherwise we receive null. That's why I used a recursive function, to know when we have to stop requesting partitions and calculate max batchId in one segment and then in the whole table. Using these two level iteration we make sure we load all partitions.

File: bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/metrics/MetricFactory.scala [hunk 5]
```diff
@@ -1,56 +1,33 @@
-/*
```
Comment (denis-pyshev) [inline bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/metrics/MetricFactory.scala]: Please return copyright notice here (and in other places if applicable)

File: bandarlog/src/main/resources/application.conf [hunk 6]
```diff
@@ -82,7 +92,7 @@ bandarlogs {                                           # system object, all band
     ]
   }
 
-  presto-vertica-bandarlogs {                          # sql bandarlog name, can be any name
+  presto-vertica-glue-bandarlogs {                     # sql bandarlog name, can be any name
```
Comment (denis-pyshev) [inline bandarlog/src/main/resources/application.conf]: I'd vote to rename to `in-out-bandarlogs` to not enumerate all connectors possibilities.

File: bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/connectors/GlueConnector.scala [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    val columnsData = values
+      .flatMap(value => value.zip(columns))
+      .filter { case (value, columnName) => columnName == tableColumn }
+
+    val max = columnsData match {
+      case Nil => throw new IllegalArgumentException(s"Column $tableColumn not found in table $tableName.")
+      case nonEmptyList: List[(String, String)] =>
+        nonEmptyList.map { case (value, columnName) => value.toLong }.max
+    }
+    max
```
[the remark this answers is not in the document]
Comment (denis-pyshev) [reply bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/connectors/GlueConnector.scala]: I'd vote against that for the sake of readability

[the remark this answers is not in the document]
Comment (alexey-lipodat) [reply bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/connectors/GlueConnector.scala]: Rechecked. Agree with you, it makes sense.

File: bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/connectors/GlueConnector.scala [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    segment
+  }
+
+  /**
+    * Fetches list of names of partition columns in table
+    *
+    * @param tableName - table name
+    * @return          - list of names of partition columns
+    */
+  private def getPartitionColumns(tableName: String): List[String] = {
```
[the remark this answers is not in the document]
Comment (MariiaRa) [reply bandarlog/src/main/scala/com/aol/one/dwh/bandarlog/connectors/GlueConnector.scala]: In case there is no table, we expect EntityNotFoundException from AWS, which will be handled by Bandar-Log.

File: project/Dependencies.scala [hunk 9]
```diff
@@ -42,4 +43,5 @@ object Dependencies {
   val scalaz                = "org.scalaz"            %% "scalaz-core"                 % V.scalaz
   val scalaTest             = "org.scalatest"         %% "scalatest"                   % V.scalatest     % Test
   val mockito               = "org.mockito"           % "mockito-core"                 % V.mockito       % Test
+  val athenaDriver          = "com.syncron.amazonaws" % "simba-athena-jdbc-driver"     % V.athenaDriver
```
Comment (alexey-lipodat) [inline project/Dependencies.scala]: I see such error logs in test env:
```
8/16/2018 1:28:40 PMlog4j:ERROR A "org.apache.log4j.ConsoleAppender" object is not assignable to a "com.simba.athena.shaded.apache.log4j.Appender" variable.
8/16/2018 1:28:40 PMlog4j:ERROR The class "com.simba.athena.shaded.apache.log4j.Appender" was loaded by
8/16/2018 1:28:40 PMlog4j:ERROR [sun.misc.Launcher$AppClassLoader@74a14482] whereas object of type
8/16/2018 1:28:40 PMlog4j:ERROR "org.apache.log4j.ConsoleAppender" was loaded by [sun.misc.Launcher$AppClassLoader@74a14482].
8/16/2018 1:28:40 PMlog4j:ERROR Could not instantiate appender named "console".
```

I suppose we can exclude all unrelated stuff like `log4j`.

File: README.md [hunk 10]
```diff
@@ -94,21 +94,32 @@ sql-data-source-config {             # data source configuration id (can be any
...
   connection.timeout.ms = 60000      # connection timeout in ms (by default = 60000)
 }
+
+aws-glue-source-config {
+  region = "region"                  # aws region
+  dbname = "database"                # database name
+  access.key = "accesskey"           # access key provided by AWS account
+  secret.key = "secretaccesskey"     # secret key provided by AWS account
+  fetch.size = 10                    # the maximum number of partitions to return in a single response
+  segment.total.number = 10          # the total nummer of segments - non-overlapping region of a table's partitions. Maximum possible value - 10
```
Comment (alexey-lipodat) [inline README.md]: Please, fix in all places `nummer` -> `number`

File: infra/src/main/scala/com/aol/one/dwh/infra/config/RichConfig.scala [hunk 11]
```diff
@@ -18,10 +18,10 @@ object RichConfig {
...
     def getOptionalString(path: String): Option[String] =
       if (underlying.hasPath(path) && !underlying.getIsNull(path)) {
-      Some(underlying.getString(path))
-    } else {
-      None
-    }
+        Some(underlying.getString(path))
+      } else {
+        None
+      }
```
[the remark this answers is not in the document]
Comment (alexey-lipodat) [reply infra/src/main/scala/com/aol/one/dwh/infra/config/RichConfig.scala]: Please, fix this comment ^

File: project/Dependencies.scala [hunk 12]
```diff
@@ -33,7 +34,7 @@ object Dependencies {
   val consulClient          = "com.ecwid.consul"      % "consul-api"                   % V.consulClient
   val hikariPool            = "com.zaxxer"            % "HikariCP"                     % V.hikariPool
   val slf4j                 = "org.slf4j"             % "slf4j-api"                    % V.slf4j
-  val log4j                 = "org.slf4j"             % "slf4j-log4j12"                % V.slf4j
+  val log4j                 = "org.slf4j"             % "slf4j-log4j12slf"                % V.slf4j
```
Comment (alexey-lipodat) [inline project/Dependencies.scala]: Why we changed it?

**Comment by denis-pyshev** [comment]
Tested.
